### DISCUSSION TITLE
Possible issue in Todos example project

### DISCUSSION THREAD

**Discussion by jestspoko** [pointfreeco/swift-composable-architecture#1690] [General]
I think there might be  an issue in Todos example. When trying to interact with row in "completed" section (deleting item in this case), item with different index is being deleted. This is probably because "all" section is powered by "todos" array while "completed" view is using `filteredTodos` which is just a slice of original `todos` database.

This issue also could appear if we decide to filter out completed todos right after completion. This would also lead to a issue with `todos` indexes. 

Possible solution would be to have separate arrays with filtered todos (similar to: https://github.com/pointfreeco/swift-composable-architecture/discussions/1627 ) but it will require a lot of moving/removing  which just feels wrong.  
Any ideas how to solve this puzzle?

https://user-images.githubusercontent.com/411102/204061271-b7421f82-d0b2-41b5-958f-900c1a15e0c7.mp4

**Comment by mbrandonw** [comment]
Hi @jestspoko, good catch!

Here's one potential solution:

```swift
case let .delete(indexSet):
  let filteredTodos = state.filteredTodos
  for index in indexSet {
    state.todos.remove(id: filteredTodos[index].id)
  }
  return .none
```

It's a little extra work because we have to compute the filtered todos again, but it's not so bad. A full solution probably would beef up the data structure holding todos so that it could more efficiently access _views_ into the data, but that may not be super appropriate for such a simple demo that's just meant to convey some basic TCA ideas.

We'll think about it a bit more to see how we want to address this.

**Reply by mbrandonw** [reply]
Thanks again for bringing this up @jestspoko! I have just pushed a fix here: https://github.com/pointfreeco/swift-composable-architecture/pull/1696/
